### ISSUE TITLE
Buttons when disabled have unsightly dark grey backgrounds

### ISSUE THREAD

**Issue by kimfaint** [kivymd/KivyMD#1097]
### Description of the Bug

Take the following three buttons: `MDIconButton`, `MDFlatButton`, `MDRaisedButton`:
![kivymd_button_enabled](https://user-images.githubusercontent.com/1551809/136664373-a9126f9c-8ca4-4d60-aa21-532ff0179af5.png)

In KivyMD 0.104.1 when these buttons are disabled they appear like:
![kivymd_button_disabled_0 104 1](https://user-images.githubusercontent.com/1551809/136664244-d3e21257-ecfc-43b3-8cd4-e3df69e354f6.png)

In KivyMD 0.104.2 and the current master, when these buttons are disabled they appear like:
![kivymd_button_disabled_0 104 2](https://user-images.githubusercontent.com/1551809/136664239-6a872b04-88bf-4b68-bde6-b787f7a5f918.png)

### Code and Logs

The following code was used to generate the screenshots.  The widgets in the second column are given `md_bg_color_disabled: [0,0,0,0]` as a workaround.  Alternative workaround on line 51, if un-commented it will monkey patch `BaseButton.md_bg_color_disabled = [0,0,0,0]` globally.

```python
from kivy.lang import Builder

from kivymd.app import MDApp
from kivymd.uix.button import BaseButton

KV = '''
MDGridLayout:
    cols: 2
    pos_hint: {'top': 1}
    padding: 50
    spacing: 50
    size_hint: None, None

    MDIconButton:
        icon: "language-python"
        disabled: sw.active

    MDIconButton:
        icon: "language-python"
        disabled: sw.active
        md_bg_color_disabled: [0,0,0,0]

    MDFlatButton:
        text: "MDFlatButton"
        disabled: sw.active

    MDFlatButton:
        text: "MDFlatButton"
        disabled: sw.active
        md_bg_color_disabled: [0,0,0,0]

    MDRaisedButton:
        text: 'MDRaisedButton'
        disabled: sw.active

    MDRaisedButton:
        text: 'MDRaisedButton'
        disabled: sw.active
        md_bg_color_disabled: [0,0,0,0]

    MDLabel:
        text: 'Disabled'

    MDSwitch:
        id: sw
        active: False
'''

class Example(MDApp):
    def build(self):
        #BaseButton.md_bg_color_disabled = [0,0,0,0]
        return Builder.load_string(KV)

Example().run()
```

### Screenshots
![kivymd_button_disabled_0 104 2_workaround](https://user-images.githubusercontent.com/1551809/136664737-75bd6699-4781-48e8-b57c-63acaae3e66c.png)

### Versions

* OS: Ubuntu 21.04 and Android 10 
* Python: 3.9.5 and 3.8.1
* Kivy: 2.0.0
* KivyMD: 0.104.2 and 9fc99af

**Comment by HeaTTheatR** [comment]
@kimfaint 
![Снимок экрана 2021-10-18 в 23 23 57](https://user-images.githubusercontent.com/16930280/137801515-6cda5bcb-73d1-499b-a7c4-f5370db4f388.png)

![Снимок экрана 2021-10-18 в 23 24 43](https://user-images.githubusercontent.com/16930280/137801607-569e9434-ee23-48f6-8ceb-6046ddb0baf8.png)

**Comment by gbm001** [comment]
Glad you obviously agree that the current inconsistent and overly-dark disabled buttons looks nothing like the subtle light grey backgrounds of the Material Design spec :)

So differences from the Spec to start with:
- for icon buttons which have no background to begin with, they should not gain one when disabled (whereas icon buttons are gaining a medium grey circular background)
- there are no examples of disabled text buttons without backgrounds in the spec, but it would be inconsistent for them to gain one (whereas flat buttons are gaining a medium grey disabled background)
- for buttons which start with a background, they gain a consistent light grey background (whereas raised buttons gain a background so dark it obscures the text of the buttons)

The KivyMD 0.104.1 behaviour is consistent with the Material Design spec (as far as stated); the KivyMD 0.104.2 behaviour is not.

**Comment by HeaTTheatR** [comment]
@gbm001 We do not support KivyMD version 0.104.2

**Comment by gbm001** [comment]
Fair enough - which version is supported, and I assume that means it is fixed? (in which case this issue can be closed; it was still open when I hit the same problem)

I think the problem is that the disabled_hint_text_color, used to set the md_bg_color_disabled, is set to 0.38 opacity of black in theming.py via _get_disabled_hint_text_color - whereas the correct opacity is 0.38 of the *active state* plus (where appropriate) colour changes (which would normally be changing to greyscale).

**Comment by HeaTTheatR** [comment]
@gbm001 If you know how to fix it - welcome to contributors :)

**Comment by gbm001** [comment]
I think the thing to do is set the md_bg_color_disabled in update_md_bg_color based on a greyscale version of md_bg_color at 38% opacity (or something like that. I will have a play.

**Comment by gbm001** [comment]
After eventually knocking something that works (by calculating the 38% opacity of whatever the background color is) and then realising that you should instead just pre-calculate 38% of the greyscale version of primary_color, I have decided not to submit it as a pull request because I don't really understand it still.

The whole thing is incredibly difficult to sort out because of the nightmarish hierarchy of button classes. For example:
MDFillRoundFlatIconButton (a filled, round class) inherits from
MDRoundFlatIconButton (a non-filled, round class) which inherits from
MDRoundFlatButton (a non-filled, round class) which inherits from
MDFlatButton (a non-filled, no-border class) which inherits from
BaseFlatButton which inherits from
BaseRectangularButton which inherits from
BaseButton (which finally inherits from ButtonBehaviour in Kivy).

That's crazily deep inheritance; it really needs refactoring with mix-in classes or something. I'll have a go at that rather than my hideous hacky 'it works so hey' approach to background colours.

**Comment by HeaTTheatR** [comment]
@gbm001 Try it :)

**Comment by gbm001** [comment]
#1139 

A first draft of a refactoring...

**Comment by gbm001** [comment]
Should all be fixed now in #1139; this bug can probably be closed?
